cleverhans/attacks.py

```diff
@@ -1,62 +1,233 @@
-    :param predictions: the model's symbolic output (linear output,
-        pre-softmax)
-    :param sample: (1 x 1 x img_rows x img_cols) numpy array with sample input
-    :param target: target class for input sample
-    :param theta: delta for each feature adjustment
-    :param gamma: a float between 0 - 1 indicating the maximum distortion
-        percentage
-    :param increase: boolean; true if we are increasing pixels, false otherwise
-    :param back: switch between TensorFlow ('tf') and
-                Theano ('th') implementation
-    :param clip_min: optional parameter that can be used to set a minimum
-                    value for components of the example returned
-    :param clip_max: optional parameter that can be used to set a maximum
-                    value for components of the example returned
-    :return: an adversarial sample
-    """
-    if back == 'tf':
-        # Compute Jacobian-based saliency map attack using TensorFlow
-        from .attacks_tf import jsma_tf
-        return jsma_tf(sess, x, predictions, grads, sample, target, theta,
-                       gamma, increase, clip_min, clip_max)
-    elif back == 'th':
-        raise NotImplementedError("Theano jsma not implemented.")
+class Attack:
+    """
+    Abstract base class for all attack classes.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, x, pred, y=None, backend='tf', clip_min=None, clip_max=None):
+        """
+        :param x: A placeholder for the model inputs.
+        :param pred: The model's symbolic output.
+        :param y: A placeholder for the model inputs. Only provide this parameter if
+                you'd like to use the true labels when crafting adversarial samples.
+                Otherwise, model predictions are used as labels to avoid the label
+                leaking effect. Default is None.
+        :param backend: A string indicating which backend to use. Must be either
+                        'tf' or 'th'. Default is 'tf'.
+        :param clip_min: Optional float parameter that can be used to set a minimum
+                        value for components of the example returned.
+        :param clip_max: Optional float parameter that can be used to set a maximum
+                        value for components of the example returned.
+        """
+        assert backend == 'tf' or backend == 'th'
+        self.x = x
+        self.pred = pred
+        self.y = y
+        self.backend = backend
+        self.clip_min = clip_min
+        self.clip_max = clip_max
+
+    def generate_symbolic(self):
+        """
+        Generate symbolic graph for adversarial samples and return. This
+        method should be overwritten in any child class which has a
+        symbolic implementation available.
+        """
+        raise NotImplementedError("The symbolic version of this attack "
+                                  "is not currently implemented.")
+
+    @abstractmethod
+    def generate_numpy(self, sess):
```

Comment:
Yes, as you mentioned we can always make non-targeted attack into targeted by choosing target class randomly. And vice versa we can make targeted attack into non-targeted by ignoring target class.

For example, iterative target class attack from one of my papers (https://arxiv.org/abs/1607.02533) always need target class, however in my experiments I made it non-targeted by using least likely class as target class.
Note: I'm not advocating to use least likely class or to use random class, it's just an example of what I did in the past.

So code may look like following:

```
class FastGradientMethod(Attack):
  def generate_numpy(X, Y=None, sess=None, batch_size=128, target_class=None):
    # FGSM is non-targeted attack, thus we just ignore target_class argument
    del target_class
    # ...

class SaliencyMapMethod(Attack):
  def generate_numpy(X, Y=None, sess=None, batch_size=128, target_class=None):
    # SaliencyMapMethod always require target class, so make it up if it's none
    if not target_class:
      target_class = ...  # make up target class
    # ...
  
```


Another alternative is to throw an exception when targeted attack called without target class. But this way it would be harder to use methods interchangeably. So I would recommend against this option.